app/models/reduction_fetcher.rb

```diff
@@ -35,9 +35,23 @@ def reduce_by_subject?
     @topic == :reduce_by_subject
   end
 
-  def where(query)
-    return @subject_reductions.where(query.except(:user_id)) if reduce_by_subject?
-    return @user_reductions.where(query.except(:subject_id)) if reduce_by_user?
+  def where(selector, in_place: false)
+    if in_place
+      return locate_in_place(selector.except(:user_id), @subject_reductions, SubjectReduction) if reduce_by_subject?
+      return locate_in_place(selector.except(:subject_id), @user_reductions, UserReduction) if reduce_by_user?
+    else
+      return @subject_reductions.where(selector.except(:user_id)).first_or_initialize if reduce_by_subject?
+      return @user_reductions.where(selector.except(:subject_id)).first_or_initialize if reduce_by_user?
+    end
+  end
+
+  def locate_in_place(selector, relation, factory)
+    match = relation.to_a.map(&:serializable_hash).select{ |record| key_match(record, selector) }
```

Comment:
I'm not sure what this method is doing as it seems to have a few responsibilities. Could it be broken apart? Perhaps inlined or split up and called by the `where` `in_place` method from the boolean param arg comment?

Minor: Does relation not respond to `each` ? might save a type conversion `to_a` if it does.

Reply:
The reduction fetcher, in the case of `running_reduction` mode, prefetches all reductions into the model so that we can detect object versioning problems that would oblige us to redo the reduction because our inputs are outdated. Previously, I just called `#where` on the prefetched collection, but @marten demonstrated that instead of using the prefetched objects, this simply issued another DB query, which was not the expected or correct behavior.

`#each` returns the original collection while producing side effects, while `#map` returns the collection with the predicate applied, so it's not really substitutable here.

Reply:
This has been merged with the `#retrieve_in_place` method for clarity

Reply:
Minor - if a method is enumerable / responds to `each` you can apply a map on it. `relation.to_a.map`, my question was why the enumerable has to be cast to an array before applying the map.